### ISSUE TITLE
Execution blocked for app data files when targeting API >= 29 (Android 10 / Q)

### ISSUE THREAD

**Issue by JeromeDeBretagne** [JeromeDeBretagne/erlanglauncher#2]
https://developer.android.com/about/versions/10/behavior-changes-10#execute-permission

When targeting API 29 (Android 10 / Q) or above, it is not possible anymore to have execute permission for files stored within the app's home directory (data), which is exactly where the Erlang runtime is currently unpacked.

This significant change is being discussed by various projects, in termux/termux-app#1072 for instance. The recommended and future-proof way is to extract program binaries into the application's native `lib` directory (with `android:extractNativeLibs=true`), where files can still be executed but are stored read-only for improved security.

The tricky part for Erlang/OTP is that the app `lib` directory has a random path created at installation/update time, so the Erlang runtime will have to be modified to become fully independent of its installation path to keep working on Android in the long run.

This Android 10 modification was introduced in commit: https://android-review.googlesource.com/c/platform/system/sepolicy/+/804149 and was then confirmed officially by Google here: https://issuetracker.google.com/issues/128554619

**Comment by JeromeDeBretagne** [comment]
Here are corresponding error messages possible in the logs, for reference:
```
Cannot run program xyz in directory abc error=13, Permission denied
avc: denied { execute_no_trans } for path="..."
```

**Comment by D4no0** [comment]
This is a painful problem, I would suggest to try to look into how it is possible to build a .so native library for android, it should be possible however not easy. A good starting point is to understand how the system is built for windows as stated in this issue: https://bugs.erlang.org/browse/ERL-1292?filter=-2

**Comment by JeromeDeBretagne** [comment]
Using a .so native library would be a way to solve this issue, but it would require to change the Erlang build system as done for Windows. Possible but not easy at all indeed, @D4no0 .

I'm considering another (simpler?) way at the moment which would only require to improve the erl script (and a few other ones) so that it can become path independent and thus relocatable, as proposed in this issue :
https://bugs.erlang.org/browse/ERL-1323

**Comment by D4no0** [comment]
The improvement you mention is a good thing by itself. The only downside I see is that this might take probably a very long time to be adopted officially, since it should be backward compatible with older versions.

The next problem is that Android system will get more secured by the day and then you will be forced to make .so libraries someday.

I will look into how the system currently builds on windows and try to create scripts for a .so library, the biggest downside is that there is no clear documentation how that was done, beside a single wiki page.

**Comment by D4no0** [comment]
Wow nice! I suppose the application name still remains hardcoded?

**Comment by JeromeDeBretagne** [comment]
A big "Thanks" to the [Termux](https://github.com/termux/termux-app) community, I've taken my inspiration from the discussions at https://github.com/termux/termux-app/issues/1072 and from commit [f6c3b6f](https://github.com/termux/termux-app/commit/f6c3b6f38aa086cd9ca50df86b1c782ed869e0cc) from @imprakharshukla in particular in the [android-10 ](https://github.com/termux/termux-app/tree/android-10) branch.

@D4no0 It took me a while but I've finally implemented a first solution for this limitation introduced in recent versions of Android. It seems to be working well on a wide range of Android versions (I've tested from Android 4.4 to 11).

Now I need to investigate how to make the code compatible with multilple users, and that's when I think I'll need a fix for issue [ERL-1323](https://bugs.erlang.org/browse/ERL-1323)

**Comment by JeromeDeBretagne** [comment]
> Wow nice! I suppose the application name still remains hardcoded?

@D4no0 indeed, but I've moved the project to a more usual "app" directory already. So far, I've not looked at other clean-ups, I'm still focusing on key big issues, I consider this project mainly as a reference sample code otherwise

**Comment by D4no0** [comment]
This is great!
Have you tried running the install script from the application itself, passing the path at runtime? In theory it should work.

**Comment by JeromeDeBretagne** [comment]
My current understanding is that the ./Install script approach won't work, as the scripts are now installed in a purely read-only directory, so it's not possible to pass the path at runtime to udpate the scripts once (as expected by the Erlang install flow).

But I have a few other ideas already, either using environment variables (it should be quite simple I hope) or auto-detection of the runtime installation location (trickier but the approach I would prefer) :-)

**Comment by D4no0** [comment]
That is great, if you manage to implement this, then I can start working on porting the scenic render pipeline to android and hopefully in the future to be able to develop android applications in elixir!!!

**Comment by 4ntoine** [comment]
Hey, guys. Can you briefly explain how to solved it? Similar problem - https://stackoverflow.com/questions/64786837/what-path-to-put-executable-to-run-on-android-29

I'm not sure i got all the details but after looking into the changes i can see some symlinking and preparations on build step + some runtime changes.

**Comment by JeromeDeBretagne** [comment]
Hey @4ntoine,

I've tried to document it already the best I could in the commit message, let me copy the explanation here:

> All native program binaries and shell scripts are now extracted into the application's native lib directory (in /data/app/the.app.dir/lib/<abi>). This is indeed the only remaining location where bundled files can still get the execute permission when targeting API level 29 or above on devices running Android 10 and higher.
> 
> To do so, all executable files are placed in the jniLibs/<abi> subdirectory (used for JNI native libraries usually) and renamed into lib___.so which is the naming convention expected by Android for libraries. This is done via a Gradle build script in build.gradle. All such lib___.so files are automatically copied within the APK file itself during the build, into the lib/<abi> directory.
> 
> In the Android Manifest file, by setting the following attribute:
>    android:extractNativeLibs="true"
> these lib___.so files are then extracted at installation time in the right native lib directory, with support for execute permission but set read-only for tightened security.
> 
> This first implementation is strongly inspired from Termux, taking a simple approach: all files from the Erlang runtime, and not only the executable ones, are extracted in the lib dir at installation time as explained above. The usual Erlang runtime file stucture is then fully recreated using symlinks in the home (data) directory, during the app initialization on first run.

The additional information is that during the preparation at build time, a file is created (libmappings.so) to store the mapping between the renamed lib___.so files and their original filenames, so that the symlinks can be created to point to the right files.

I hope this helps a bit to get a first understanding of the principle used in this implementation.

**Comment by JeromeDeBretagne** [comment]
@D4no0 so I have now a working solution running locally to support properly multiple users on Android, as I had imagined with the simpler approach using an environment variable :-)

I've created the following pull request https://github.com/erlang/otp/pull/2863 to propose this approach as an official solution to the Erlang/OTP project, let's see what the team thinks about it and if they find it acceptable. If accepted upstream, I'll be able to update this project quickly with my local changes, using the official approach.

**Comment by 4ntoine** [comment]
@JeromeDeBretagne Thank you so much for summarizing and explaining it!

> all files from the Erlang runtime, and not only the executable ones, are extracted in the lib dir

Do you know what is the motivation to extract **all the files** into `/lib`, why not just extract the executables and having the actual files in app home dir (so only executables are actually symlinks to the ones in /lib and the others are actual files?).

My main concern is about compression ratio: that since .apk are actually zip files the [android] compression ratio is much worse compared to something compressed with 7zip for instance and bundled right in app raw resources? So only executables can be handled as shared libs (with zip compression ratio) and the rest files could be compessed with 7zip, bundle in app raw resources and extracted into app home dir with much better compression ratio.

**Comment by JeromeDeBretagne** [comment]
> Do you know what is the motivation to extract **all the files** into `/lib`, why not just extract the executables and having the actual files in app home dir (so only executables are actually symlinks to the ones in /lib and the others are actual files?).

That's an easy one, my motivation was : the ease of implementation, as I wasn't 100% sure it would work when I started this investigation.

> So only executables can be handled as shared libs (with zip compression ratio) and the rest files could be compessed with 7zip, bundle in app raw resources and extracted into app home dir with much better compression ratio.

Your proposition is perfectly possible, so you should totally go for it! That was a 2nd option I had in mind (and that I still have in mind in fact) hence this part of my commit message  :-)
> This first implementation ...

**Comment by 4ntoine** [comment]
Thanks, @JeromeDeBretagne !

**Comment by D4no0** [comment]
> @JeromeDeBretagne Thank you so much for summarizing and explaining it!
> 
> > all files from the Erlang runtime, and not only the executable ones, are extracted in the lib dir
> 
> Do you know what is the motivation to extract **all the files** into `/lib`, why not just extract the executables and having the actual files in app home dir (so only executables are actually symlinks to the ones in /lib and the others are actual files?).
> 
> My main concern is about compression ratio: that since .apk are actually zip files the [android] compression ratio is much worse compared to something compressed with 7zip for instance and bundled right in app raw resources? So only executables can be handled as shared libs (with zip compression ratio) and the rest files could be compessed with 7zip, bundle in app raw resources and extracted into app home dir with much better compression ratio.

There are some resources that require to be compressed however I wouldn't rely that much on runtime decompression since it takes a lot of time. For example currently launching beam on android takes a good few seconds, if we were to use for example releases configs provided by elixir it would be even longer (starts the beam, reloads configs then restarts the beam).

**Comment by D4no0** [comment]
> My current understanding is that the ./Install script approach won't work, as the scripts are now installed in a purely read-only directory, so it's not possible to pass the path at runtime to udpate the scripts once (as expected by the Erlang install flow).
> 
> But I have a few other ideas already, either using environment variables (it should be quite simple I hope) or auto-detection of the runtime installation location (trickier but the approach I would prefer) :-)

What about running install in a gradle task, in theory this way you are configuring the paths before deploying on device?

**Comment by JeromeDeBretagne** [comment]
> What about running install in a gradle task, in theory this way you are configuring the paths before deploying on device?

The newer implementation from earlier this week doesn't even need this trick, the installation path for the main user is hard-coded and it mostly works with the following 2 major limitations:
1/ this is not future-proof as this path is not documented, shouldn't be used as-is and could change on next Android revisions  
2/ it doesn't support multiple-user setup (tablets or Chromebook with several accounts, work profiles on smartphones, etc.)  

To be more explicit about the second limitation, here is how a (rooted) device with 2 users looks like:

```
# ls /data/user                                                                                                                                                                            
0  10
```
The main user has its app storage reachable equally via /data/data or /data/user/0 while other users while have their app data stored in /data/user/ID. The Erlang Runtime file structure is re-created in two different locations for each user:

```
# ls -l /data/data/com.isotope11.erlanglauncher/files/erlang/                                                                                                                            
total 18
lrwxrwxrwx 1 u0_a260 u0_a260  114 2020-11-11 15:37 Install -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib100.so
drwx------ 2 u0_a260 u0_a260 3488 2020-11-11 15:37 bin
drwx------ 6 u0_a260 u0_a260 3488 2020-11-08 23:10 erts-11.0.2
drwx------ 8 u0_a260 u0_a260 3488 2020-11-08 23:10 lib
drwx------ 2 u0_a260 u0_a260 3488 2020-11-11 15:37 misc
drwx------ 3 u0_a260 u0_a260 3488 2020-11-11 15:37 releases
drwx------ 4 u0_a260 u0_a260 3488 2020-11-08 23:10 usr

# ls -l /data/user/10/com.isotope11.erlanglauncher/files/erlang/                                                                                                                           
total 18
lrwxrwxrwx 1 u10_a260 u10_a260  114 2020-11-14 10:28 Install -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib100.so
drwx------ 2 u10_a260 u10_a260 3488 2020-11-14 10:28 bin
drwx------ 6 u10_a260 u10_a260 3488 2020-11-11 14:48 erts-11.0.2
drwx------ 8 u10_a260 u10_a260 3488 2020-11-11 14:48 lib
drwx------ 2 u10_a260 u10_a260 3488 2020-11-14 10:28 misc
drwx------ 3 u10_a260 u10_a260 3488 2020-11-14 10:28 releases
drwx------ 4 u10_a260 u10_a260 3488 2020-11-11 14:48 usr
```
and here is the content of the erl script, which is the same for the 2 users:

```
# cat /data/data/com.isotope11.erlanglauncher/files/erlang/bin/erl                                                                                                                         
#!/system/bin/sh
#
if [ -z "$ROOTDIR" ]
then
    ROOTDIR="/data/data/com.isotope11.erlanglauncher/files/erlang"
fi
BINDIR=$ROOTDIR/erts-11.0.2/bin
EMU=beam
PROGNAME=erl
export EMU
export ROOTDIR
export BINDIR
export PROGNAME
exec "$BINDIR/erlexec" ${1+"$@"}
```

ROOTDIR so far has not been configurable at runtime and it must be an absolute path so it could only point to one directory, here /data/data/com.isotope11.erlanglauncher/files/erlang . However, this fails for any other user which would expect the script to point to  /data/user/ID/com.isotope11.erlanglauncher/files/erlang instead. And as a reminder, the script must now be ... read-only to be executable on Android 10+ with targetSdkVersion 29 and above.

Using the updated erl script above, as I've proposed in erlang/otp#2863, ROOTDIR can be passed as an environment variable too and can point to the right absolute path for different users, there is no need for a hardcoded value anymore.

Finally, for anyone willing to better understand the notion of symlinks I've mentioned prievously, here is what it looks like for all the files re-created in the bin directory as one example, all the files being in lib/arm64 using the lib___.so format  :
```
# ls -1 /data/data/com.isotope11.erlanglauncher/files/erlang/bin/                                                                                                                          
ct_run
dialyzer
epmd
erl
erl_call
erlc
escript
no_dot_erlang.boot
run_erl
start
start.boot
start.script
start_clean.boot
start_erl
start_sasl.boot
to_erl
typer


# ls -l /data/data/com.isotope11.erlanglauncher/files/erlang/bin/
total 0
lrwxrwxrwx 1 u0_a260 u0_a260 114 2020-11-14 10:50 ct_run -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib957.so
lrwxrwxrwx 1 u0_a260 u0_a260 114 2020-11-14 10:50 dialyzer -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib955.so
lrwxrwxrwx 1 u0_a260 u0_a260 114 2020-11-14 10:50 epmd -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib952.so
lrwxrwxrwx 1 u0_a260 u0_a260 114 2020-11-14 10:50 erl -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib950.so
lrwxrwxrwx 1 u0_a260 u0_a260 114 2020-11-14 10:50 erl_call -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib954.so
lrwxrwxrwx 1 u0_a260 u0_a260 114 2020-11-14 10:50 erlc -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib953.so
lrwxrwxrwx 1 u0_a260 u0_a260 114 2020-11-14 10:50 escript -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib958.so
lrwxrwxrwx 1 u0_a260 u0_a260 114 2020-11-14 10:50 no_dot_erlang.boot -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib964.so
lrwxrwxrwx 1 u0_a260 u0_a260 114 2020-11-14 10:50 run_erl -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib960.so
lrwxrwxrwx 1 u0_a260 u0_a260 114 2020-11-14 10:50 start -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib951.so
lrwxrwxrwx 1 u0_a260 u0_a260 114 2020-11-14 10:50 start.boot -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib965.so
lrwxrwxrwx 1 u0_a260 u0_a260 114 2020-11-14 10:50 start.script -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib966.so
lrwxrwxrwx 1 u0_a260 u0_a260 114 2020-11-14 10:50 start_clean.boot -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib962.so
lrwxrwxrwx 1 u0_a260 u0_a260 114 2020-11-14 10:50 start_erl -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib959.so
lrwxrwxrwx 1 u0_a260 u0_a260 114 2020-11-14 10:50 start_sasl.boot -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib963.so
lrwxrwxrwx 1 u0_a260 u0_a260 114 2020-11-14 10:50 to_erl -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib961.so
lrwxrwxrwx 1 u0_a260 u0_a260 114 2020-11-14 10:50 typer -> /data/app/~~FOY1IYJ25vMvn7r1G_NsaA==/com.isotope11.erlanglauncher-d4kn8otlknlRXe1X7vNC2A==/lib/arm64/lib956.so
```

**Comment by JeromeDeBretagne** [comment]
> @D4no0 so I have now a working solution running locally to support properly multiple users on Android, as I had imagined with the simpler approach using an environment variable :-)
> 
> I've created the following pull request [erlang/otp#2863](https://github.com/erlang/otp/pull/2863) to propose this approach as an official solution to the Erlang/OTP project, let's see what the team thinks about it and if they find it acceptable. If accepted upstream, I'll be able to update this project quickly with my local changes, using the official approach.

For reference, my proposed pull request  [erlang/otp#2863](https://github.com/erlang/otp/pull/2863) has been accepted upstream and then updated in a follow-up fix (erlang/otp@5475a9e) so the supported environment variable is **ERL_ROOTDIR** in the end.

As planned, I've updated the project to support properly multiple users on Android in commit JeromeDeBretagne/erlanglauncher@d64e700.

As a positive side effect, this sample code should now be independent of the application package name as there is no more hard-coded paths.
